core/engine.go

```diff
@@ -109,10 +111,47 @@ func NewEngine(
 	return e, nil
 }
 
+// StartOutputs spins up all configured outputs, giving the thresholds to any
+// that can accept them. And if some output fails, stop the already started
+// ones. This may take some time, since some outpts make initial network
+// requests to set up whatever remote services are going to listen to them.
+//
+// TODO: this doesn't really need to be in the Engine, so take it out?
+func (e *Engine) StartOutputs() error {
+	e.logger.Debugf("Starting %d outputs...", len(e.outputs))
+	for i, out := range e.outputs {
+		if thresholdOut, ok := out.(output.WithThresholds); ok {
+			thresholdOut.SetThresholds(e.thresholds)
+		}
+
+		if err := out.Start(); err != nil {
+			e.stopOutputs(i)
+			return err
+		}
+	}
+	return nil
+}
+
+// StopOutputs stops all configured outputs.
+func (e *Engine) StopOutputs() {
+	e.stopOutputs(len(e.outputs))
+}
+
+func (e *Engine) stopOutputs(upToID int) {
+	e.logger.Debugf("Stopping %d outputs...", upToID)
+	for i := 0; i < upToID; i++ {
+		if err := e.outputs[i].Stop(); err != nil {
+			e.logger.WithError(err).Errorf("Stopping output %d failed", i)
```

Comment:
The problem here is that I don't have a `Name()`, I have a `Description()`, which wouldn't be suitable here :disappointed: ... To improve the error message, I either have to change the `output.Output` interface and force every output to expose a `Name()`, or I'd have to add a wrapper struct for outputs in the `Engine` that contains its type (or whole `Params`). And I can't just use a map, since users can have multiple outputs with the same type.

Reply:
> I either have to change the `output.Output` interface and force every output to expose a `Name()`

:+1: That sounds OK to me. `Description()` is too "loose" of an interface and outputs can return whatever they want there. The description should rather be assembled from different output information so that we can control how it's rendered. Maybe even `Version()` should be added, and make it no-op for built-in outputs.

Anyway, it's a minor thing, not a big deal.

Reply:
hmm :-1: for `Version()`... It would be something that only extensions would use, and it would add unnecessary complexity for writing them. Only xk6 cares about the versions for now, and if we ever start caring, I'd rather figure out some way to extract their versions from [`runtime/debug.BuildInfo`](https://golang.org/pkg/runtime/debug/#BuildInfo) than forcing extension authors to deal with built-time linker flags or bumping an in-code `Version` constant on every release of their extension.

Reply:
about `Name()`, cc @MStoykov what do you think? I'd personally rather have a wrapper struct around `output.Output` than forcing every output to have one extra helper method.

Reply:
yeah I also think it's better if we can just use the name we already have ... otherwise I would argue that we should get the name from `Name()` everywhere.

I just realized is that this makes it ... harder to reason about if you decide to have multiple outputs of the same type .... is this even supported now?

I am not certain that was supported even before that, and I am not certain is all that needed, but seems like something that can be useful especially if outputs can filter metrics on their own. So you basically can (for example) shard based on something between multiple influxdb instances ?

Reply:
> otherwise I would argue that we should get the name from Name() everywhere.

We can't, since `Name()` would be a method of an output instance, and we can't instantiate the output without giving it some valid parameters first. That's why we have the output type as a string in https://github.com/loadimpact/k6/blob/7e1bc5dcab910f0d197a63e843e5591de4c60423/output/extensions/registry.go#L56

The only way around that is to have 2 interfaces, the current `Output` and a new `OutputConstructor` which has a `Name() string` property and a `New(Params) Output` method. I don't mind that refactoring, but it would require every output to have these two interfaces, and likely an empty struct with the second one that is passed to the `Register()` function.

> I just realized is that this makes it ... harder to reason about if you decide to have multiple outputs of the same type .... is this even supported now?

Yes, it was supported even before this PR... The main problem with it is that JSON and environment variable configuration would be shared between all outputs, because they are based on the type names. Only the CLI flag config could be distinct. I've explored this issue and potential solution a bit before (https://github.com/loadimpact/k6/issues/587#issuecomment-416178562 and in #883), but I don't think it's a very urgent problem to solve.

Reply:
:+1: